readme.txt
=== Easy Digital Downloads - Coupon Importer ===
Author URI: http://www.chriscct7.com
Plugin URI: http://easydigitaldownloads.com/chris-is-awesome

Import discounts into Easy Digital Downloads via CSV

== Description ==

This is an extension for [Easy Digital Downloads](http://wordpress.org/extend/plugins/easy-digital-downloads/) that allows you to bulk import your products from a CSV file. If you wish to report a bug or a feature request for this importer, please do so on the EDD support forums.

== Installation ==

1. Activate the plugin
2. Activate your license in Plugins -> EDD CI
3. Go to Downloads > Settings > Misc and configure the CSV Import Mapping
4. Format your CSV according to the mapping in 1
5. Go to Downloads > CSV Import and upload your import file
6. Check the "Validate File Only" if you want to check if your file is valid without importing

== Changelog ==

= 1.0.0 =

* Initial release

changelog.txt
1.0: 2013-05-17

 * Initial release

1.1

 * Fixes a few bugs
 * Product requirement column now takes the post id of the download, or if multiple, seperate using pipes:
   http://www.ils.unc.edu/courses/2006_fall/inls461_001/fall06/unix/basics/images/pipeKey.png
   like: 20|32|109

includes/admin-page.php
<?php
/**
 * Admin Page
 *
 * @package     Easy Digital Downloads - Coupon Import
 * @subpackage  Admin Page
 * @copyright   Copyright (c) 2013, Chris Christoff
 * @since       1.0
*/


/**
 * Add Options Link
 *
 * Creates the admin submenu page for CSV Coupon Import.
 *
 * @access      private
 * @since       1.0
 * @return      void
*/

function edd_csv_coupon_import_menu() {
	global $edd_csv_coupon_import_page;
		$edd_csv_coupon_import_page = add_submenu_page( 'edit.php?post_type=download', __('Easy Digital Download CSV Coupon Import', 'edd'), __('Import Coupons', 'edd'), 'manage_options', 'edd-csv-coupon-import', 'edd_csv_coupon_import_page' );
}
add_action( 'admin_menu', 'edd_csv_coupon_import_menu', 10 );


/**
 * Coupon Import Page
 *
 * Shows the import settings and import button.
 *
 * @access      public
 * @since       1.0
 * @return      void
*/
function edd_csv_coupon_import_page() {
	global $edd_options;

	?>
	<div class="wrap">
		<h2><?php _e('CSV Coupon Import', 'edd'); ?></h2>
		<?php

			do_action('edd_csv_import_coupons_top');

			// The upload form
			edd_show_coupon_upload_form();

			edd_show_coupon_import_log();

			do_action('edd_coupon_import_bottom');
		?>
	</div><!--end wrap-->
	<?php
}


includes/admin-actions.php
<?php
/**
 * Admin Actions
 *
 * @package     Easy Digital Downloads - Coupon Import
 * @subpackage  Admin Actions
 * @copyright   Copyright (c) 2013, Chris Christoff
 * @since       1.0
*/


/**
 * Import Coupon File
 *
 * @access      public
 * @since       1.0
 * @return      void
*/
function edd_import_process_coupon_file() {
	global $edd_message;
	$edd_message = '';

	if ( isset( $_POST['edd-action'] ) && ( $_POST['edd-action'] == 'import_coupon_csv' ) ) {

		// check for file

		if ( empty( $_FILES ) || $_FILES['import_file']['size'] == 0 ){
			$edd_message = 'Please choose a file to import.';
			return;
		}

		edd_import_coupon_csv_file();

		$edd_message = "Import completed.  Results in Output Log below.";
	}

}
add_action( 'init', 'edd_import_process_coupon_file', 10 );



function edd_ci_license_menu() {
	add_plugins_page( 'EDD CI', 'EDD CI', 'manage_options', 'edd-ci-license', 'edd_ci_license_page' );
}
		
function edd_ci_license_page() {
	$license 	= get_option( 'edd_ci_license_key' );
	$status 	= get_option( 'edd_ci_license_status' );
	?>
			<div class="wrap">
				<h2><?php _e('EDD CI License Options'); ?></h2>
				<form method="post" action="options.php">
				
				<?php settings_fields('edd_ci_license'); ?>
			
				<table class="form-table">
					<tbody>
						<tr valign="top">	
							<th scope="row" valign="top">
								<?php _e('License Key'); ?>
							</th>
							<td>
								<input id="edd_ci_license_key" name="edd_ci_license_key" type="text" class="regular-text" value="<?php esc_attr_e( $license ); ?>" />
								<label class="description" for="edd_ci_license_key"><?php _e('Enter your license key'); ?></label>
							</td>
						</tr>
						<?php if( false !== $license ) { ?>
						<tr valign="top">	
							<th scope="row" valign="top">
								<?php _e('Activate License'); ?>
							</th>
							<td>
								<?php if( $status !== false && $status == 'valid' ) { ?>
									<span style="color:green;"><?php _e('active'); ?></span>
									<?php wp_nonce_field( 'edd_ci_nonce', 'edd_ci_nonce' ); ?>
									<input type="submit" class="button-secondary" name="edd_license_deactivate" value="<?php _e('Deactivate License'); ?>"/>
								<?php } else {
									wp_nonce_field( 'edd_ci_nonce', 'edd_ci_nonce' ); ?>
									<input type="submit" class="button-secondary" name="edd_license_activate" value="<?php _e('Activate License'); ?>"/>
								<?php } ?>
							</td>
						</tr>
					<?php } ?>
				</tbody>
				</table>	
				<?php submit_button(); ?>
		
			</form>
		<?php
		}

		function edd_ci_register_option() {
			// creates our settings in the options table
			register_setting('edd_ci_license', 'edd_ci_license_key', 'edd_sanitize_license' );
		}

		function edd_sanitize_license( $new ) {
			$old = get_option( 'edd_ci_license_key' );
			if( $old && $old != $new ) {
				delete_option( 'edd_ci_license_status' ); // new license has been entered, so must reactivate
			}
		return $new;
		}

		function edd_ci_activate_license() {
	
			// listen for our activate button to be clicked
			if( isset( $_POST['edd_license_activate'] ) ) {

			// run a quick security check 
			if( ! check_admin_referer( 'edd_ci_nonce', 'edd_ci_nonce' ) ) 	
				return; // get out if we didn't click the Activate button

			// retrieve the license from the database
			$license = trim( get_option( 'edd_ci_license_key' ) );
			

			// data to send in our API request
			$api_params = array( 
				'edd_action'=> 'activate_license', 
				'license' 	=> $license, 
				'item_name' => urlencode( EDD_CI ) // the name of our product in EDD
			);
		
			// Call the custom API.
			$response = wp_remote_get( add_query_arg( $api_params, EDD_CI_STORE_URL ), array( 'timeout' => 15, 'sslverify' => false ) );

			// make sure the response came back okay
			if ( is_wp_error( $response ) )
				return false;

			// decode the license data
			$license_data = json_decode( wp_remote_retrieve_body( $response ) );
		
			// $license_data->license will be either "active" or "inactive"

			update_option( 'edd_ci_license_status', $license_data->license );

			}
		}

		function edd_ci_deactivate_license() {

			// listen for our activate button to be clicked
			if( isset( $_POST['edd_license_deactivate'] ) ) {

				// run a quick security check 
				if( ! check_admin_referer( 'edd_ci_nonce', 'edd_ci_nonce' ) ) 	
					return; // get out if we didn't click the Activate button

					// retrieve the license from the database
					$license = trim( get_option( 'edd_ci_license_key' ) );
			

					// data to send in our API request
				$api_params = array( 
					'edd_action'=> 'deactivate_license', 
					'license' 	=> $license, 
					'item_name' => urlencode( EDD_CI ) // the name of our product in EDD
				);
		
				// Call the custom API.
				$response = wp_remote_get( add_query_arg( $api_params, EDD_CI_STORE_URL ), array( 'timeout' => 15, 'sslverify' => false ) );

				// make sure the response came back okay
				if ( is_wp_error( $response ) )
					return false;

				// decode the license data
				$license_data = json_decode( wp_remote_retrieve_body( $response ) );
		
				// $license_data->license will be either "deactivated" or "failed"
				if( $license_data->license == 'deactivated' )
					delete_option( 'edd_ci_license_status' );

			}
		}


		function edd_ci_check_license() {

			global $wp_version;

			$license = trim( get_option( 'edd_ci_license_key' ) );
		
			$api_params = array( 
				'edd_action' => 'check_license', 
				'license' => $license, 
				'item_name' => urlencode( EDD_CI ) 
			);

			// Call the custom API.
			$response = wp_remote_get( add_query_arg( $api_params, EDD_CI_STORE_URL ), array( 'timeout' => 15, 'sslverify' => false ) );


			if ( is_wp_error( $response ) )
				return false;

			$license_data = json_decode( wp_remote_retrieve_body( $response ) );

			if( $license_data->license == 'valid' ) {
				return true;
				// this license is still valid
			} else {
				return false;
				// this license is no longer valid
			}
		}


includes/import-functions.php
<?php
/**
 * Import Functions
 *
 * @package     Easy Digital Downloads - CSV Import
 * @subpackage  Import Functions
 * @copyright   Copyright (c) 2012, Daniel Espinzoa
 * @license     http://opensource.org/licenses/gpl-2.0.php GNU Public License
 * @since       1.0
*/


/**
 * Show File Upload
 *
 * @access      public
 * @since       1.0
 * @return      void
*/

function edd_show_coupon_upload_form() {
	global $edd_import_field_mapping, $edd_import_coupon_file_columns, $edd_message;

	if ( $edd_message != '') { ?>
	<div id="" class="updated settings-error">
		<p><strong><?php echo $edd_message ?></strong></p>
	</div>
	<?php } ?>
	<h3><?php  _e('Import Coupons From A CSV File', 'edd') ?></h3>
	<p>
		<?php _e('To import coupons select a valid CSV file and press Import File.  Results  of the import will be displayed in the log area below.', 'edd') ?>
	</p>
	<p>
		<?php
		$link = '<a href="' . get_admin_url() . 'edit.php?post_type=download&page=edd-settings&tab=misc" >' . __('Coupon Import Mapping', 'edd') . '</a>';
		echo sprintf( __('The mapping of coupon fields to CSV columns can be adjusted in the %s section.','edd'), $link );
		?>
	</p>
	<form method="post" enctype="multipart/form-data">
		<table class="form-table">
		<tbody>
			<tr valign="top">
				<th scope="row"><?php _e('Import File','edd') ?></th>
				<td><input type="file" id="import_file" name="import_file" /></td>
			</tr>
			<tr valign="top">
				<th scope="row"><?php _e('Validate File Only','edd') ?></th>
				<td><input type="checkbox" id="check_file" name="check_file" /> <?php _e('Check file, but do not save coupons to the database.','edd') ?></td>
			</tr>
		</tbody>
		</table>
		<?php
			$edd_generate_mapping_nonce = wp_create_nonce('edd_generate_mapping');
		?>
		<input type="hidden" name="_wpnonce" value="<?php echo $edd_generate_mapping_nonce  ?>" />
		<input type="hidden" name="edd-action" value="import_coupon_csv" />
		<p>
			<input type="submit" name="" id="" class="button-secondary action" value="<?php _e('Import File','edd'); ?>" />
		</p>
	</form>

	<?php
}


/**
 * Show Import Log
 *
 * @access      public
 * @param 		edd_log_file - The log file to display
 * @since       1.0
 * @return      void
*/

function edd_show_coupon_import_log() {
	global $edd_log_file;
	?>
	<strong>Output Log</strong><br/>
	<textarea style="width: 650px; height: 150px" ><?php echo $edd_log_file ?></textarea>
	<?php
}


/**
 * Import CSV File
 *
 * @access      public
 * @since       1.0
 * @return      void
*/
function edd_import_coupon_csv_file() {
	global $edd_options, $edd_log_file;

	$file_name = $_FILES['import_file']['name'];
	$edd_log_file = sprintf( __('Starting import of file: %s' , 'edd'), $file_name ) . "\n";

	if ( is_file($_FILES['import_file']['tmp_name'])) {
		if ( ($handle = fopen( $_FILES['import_file']['tmp_name'], "r")) !== FALSE ) {

			$row_ctr=0;
			while (($line = fgetcsv($handle, 1000,",")) !== FALSE ) {

				$row_ctr++;

				// skip header row
				if ( $row_ctr > 1 ) {

					// get the current post's data from the mapped column
					$edd_discount_name				= isset( $edd_options['edd_import_discount_name'] ) 			 ? $line[ $edd_options['edd_import_discount_name'] ]         : $line[0];
					$edd_discount_code				= isset( $edd_options['edd_import_discount_code'] ) 			 ? $line[ $edd_options['edd_import_discount_code'] ]         : $line[1];
					$edd_discount_type				= isset( $edd_options['edd_import_discount_type'] ) 			 ? $line[ $edd_options['edd_import_discount_type'] ]         : $line[2];
					$edd_discount_start				= isset( $edd_options['edd_import_discount_start'] ) 			 ? $line[ $edd_options['edd_import_discount_start'] ]        : $line[3];					
					$edd_discount_end				= isset( $edd_options['edd_import_discount_end'] ) 				 ? $line[ $edd_options['edd_import_discount_end'] ]          : $line[4];
					$edd_discount_status			= isset( $edd_options['edd_import_discount_status'] ) 			 ? $line[ $edd_options['edd_import_discount_status'] ]       : $line[5];
					$edd_discount_uses				= isset( $edd_options['edd_import_discount_uses'] ) 			 ? $line[ $edd_options['edd_import_discount_uses'] ]     	 : $line[6];
					$edd_discount_max_uses			= isset( $edd_options['edd_import_discount_max_uses'] ) 		 ? $line[ $edd_options['edd_import_discount_max_uses'] ]     : $line[7];
					$edd_discount_amount			= isset( $edd_options['edd_import_discount_amount'] ) 			 ? $line[ $edd_options['edd_import_discount_amount'] ] 		 : $line[8];
					$edd_discount_min_price			= isset( $edd_options['edd_import_discount_min_price']) 		 ? $line[ $edd_options['edd_import_discount_min_price'] ]    : $line[9];
					$edd_discount_product_reqs		= isset( $edd_options['edd_import_discount_product_reqs']) 		 ? $line[ $edd_options['edd_import_discount_product_reqs'] ]    : $line[10];		
					$edd_discount_product_condition	= isset( $edd_options['edd_import_discount_product_condition'])  ? $line[ $edd_options['edd_import_discount_product_condition'] ]    : $line[11];
					$edd_discount_is_not_global		= isset( $edd_options['edd_import_discount_is_not_global']) 	 ? $line[ $edd_options['edd_import_discount_is_not_global'] ]    : $line[12];
					$edd_discount_is_single_use		= isset( $edd_options['edd_import_discount_is_single_use']) 	 ? $line[ $edd_options['edd_import_discount_is_single_use'] ]    : $line[13];
					$edd_discount_id				= isset( $edd_options['edd_import_discount_id']) 				 ? $line[ $edd_options['edd_import_discount_id'] ]    : $line[14];

					
					// check for valid data
					$validation_error = false;
					$validation_msg  = '';

					// discount name  blank
					if ( 0 == strlen( $edd_discount_name ) ) {
						$validation_error = true;
						$validation_msg  .= __('Discount Name is blank.', 'edd') . "; ";
					}
					
					// discount code  blank
					if ( 0 == strlen( $edd_discount_code ) ) {
						$validation_error = true;
						$validation_msg  .= __('Discount code is blank.', 'edd') . "; ";
					}
					
					// discount type is blank
					if ( 0 == strlen( $edd_discount_type ) ) {
						$validation_error = true;
						$validation_msg  .= __('Discount type is blank.', 'edd') . "; ";
					}
					
					// discount type is invaid
					if ( ( $edd_discount_type != 'flat' ) && ( $edd_discount_type != 'percentage' ) ) {
						$validation_error = true;
						$validation_msg  .= __('Discount type is invalid.', 'edd') . "; ";
					}
					
					// discount code is blank
					if ( 0 == strlen( $edd_discount_code ) ) {
						$validation_error = true;
						$validation_msg  .= __('Discount code is blank.', 'edd') . "; ";
					}
					
					// start time validation
					if ( ! strtotime( $edd_discount_start ) ) {
						$validation_error = true;
						$validation_msg  .= sprintf(__('Date validation failed for string: %s', 'edd'), $edd_discount_start) . "; ";
					}
					
					// end time validation
					if ( ! strtotime( $edd_discount_end ) ) {
						$validation_error = true;
						$validation_msg  .= sprintf(__('Date validation failed for string: %s', 'edd'), $edd_discount_end) . "; ";
					}
					
					// set start time to correct date type
					$edd_discount_start =  date( 'm/d/Y H:i:s', strtotime( $edd_discount_start ) );
					
					// set end time to the correct date type
					$edd_discount_end =  date( 'm/d/Y H:i:s',  strtotime(  date( 'm/d/Y', strtotime( $edd_discount_end ) ) . ' 23:59:59' )  );
				
					// check status 
					$status = array( 'active', 'inactive', 'expired' );
					if ( ! in_array( $edd_discount_status, $status ) ) {
						$validation_error = true;
						$validation_msg  .= __('Status needs to be active, inactive, or pending', 'edd') . "; ";
					}
					
					// maybe more validation later


					if ( $validation_error ) {
						// report error and loop
						$edd_log_file .= sprintf( __('Error importing row %d: ', 'edd'), $row_ctr)  . $validation_msg . "\n" ;
					} else {

						if ( isset($_POST['check_file'] ) && ( $_POST['check_file'] == 'on' ) ) {

							$edd_log_file .= sprintf(__('Successfully validated row: %d', 'edd'), $row_ctr ) . "\n";

						} else {
							
							$edd_discount_product_reqs = explode("|",$edd_discount_product_reqs);
							foreach($edd_discount_product_reqs as $index => $val){
								$edd_discount_product_reqs[$index] = trim($val);  
							}
							$meta = array(
									'code'       		=> $edd_discount_code,
									'uses'       		=> $edd_discount_uses,
									'max_uses'   		=> $edd_discount_max_uses,
									'amount'     		=> $edd_discount_amount,
									'start'      		=> $edd_discount_start,
									'expiration' 		=> $edd_discount_end,
									'type'       		=> $edd_discount_type,
									'min_price'   		=> $edd_discount_min_price,
									'product_reqs'      => $edd_discount_product_reqs,
									'product_condition' => $edd_discount_product_condition,
									'is_not_global'     => $edd_discount_is_not_global,
									'is_single_use'     => $edd_discount_is_single_use
							);
						
							// Is it already made?
							if ( ( 0 == strlen( $edd_discount_id ) ) && edd_discount_exists( $edd_discount_id ) ){
							
								wp_update_post( array(
									'ID'          => $edd_discount_id,
									'post_title'  => $edd_discount_name,
									'post_status' => $edd_discount_status
								) );

								foreach( $meta as $key => $value ) {
									if ( is_string( $value ) || is_int( $value ) ){
										update_post_meta( $edd_discount_id , '_edd_discount_' . $key, strip_tags( addslashes( $value ) ) );
									}
									elseif ( is_array( $value ) ){
										update_post_meta( $edd_discount_id , '_edd_discount_' . $key, array_map( 'absint',  $value ) );										
									}
									else {
										update_post_meta( $edd_discount_id , '_edd_discount_' . $key, $value );
									}
								}
								$edd_log_file .= sprintf(__('Successfully updated row: %d', 'edd'), $row_ctr ) . "\n";
								
							}
							else {
							
								// add post
								// setup post args
								$new_download_args = array(
									'post_type' 	 => 'edd_discount',
									'post_title' 	 => $edd_discount_name,
									'post_status' 	 => $edd_discount_status,
								);
								$new_download_id = wp_insert_post( $new_download_args , true );
							

								foreach ( $meta as $key => $value ) {
									if ( is_string( $value ) || is_int( $value ) ){
										update_post_meta( $new_download_id , '_edd_discount_' . $key, strip_tags( addslashes( $value ) ) );
									}
									elseif ( is_array( $value ) ){
										update_post_meta( $new_download_id , '_edd_discount_' . $key, array_map( 'absint',  $value ) );										
									}
									else {
										update_post_meta( $new_download_id , '_edd_discount_' . $key, $value );
									}
								}
								$edd_log_file .= sprintf(__('Successfully imported row: %d', 'edd'), $row_ctr ) . "\n";
							
							}

							if ( is_wp_error( $new_download_id ) ) {

								$edd_log_file .= sprintf( __('Error importing row %d: ', 'edd'), $row_ctr) . $new_download_id->get_error_message() ;

							}

						} // if ( just validating )

					} // if ( $validation_error )

				} // if ( $row_ctr )

			} // while()

		} else {

			$edd_log_file .= __('Error opening file uploaded.','edd');
		}

	} else {

		$edd_log_file .= __('Error opening file uploaded.', 'edd');
	}


}

easy-digital-downloads-coupon-importer.php
<?php
/*
Plugin Name: Easy Digital Downloads - Coupon Importer
Plugin URI: http://easydigitaldownloads.com/chris-is-amazing
Description: Adds the ability to import coupons via a CSV file.
Author: Chris Christoff
Author URI: http://www.chriscct7.com
Version: 1.1
*/

class EDD_CI {

	private static $instance;


	/**
	 * Get active object instance
	 *
	 * @since 1.0
	 *
	 * @access public
	 * @static
	 * @return object
	 */
	public static function get_instance() {

		if ( ! self::$instance )
			self::$instance = new EDD_CI();

		return self::$instance;
	}

	/**
	 * Class constructor.  Includes constants, includes and init method.
	 *
	 * @since 1.0
	 *
	 * @access public
	 * @return void
	 */
	public function __construct() {

		define( 'EDD_CI_STORE_API_URL', 'https://easydigitaldownloads.com' );
		define( 'EDD_CI_PRODUCT_NAME', 'Coupon Importer' );
		define( 'EDD_CI_VERSION', '1.1' );
		// plugin folder url
		if ( !defined( 'EDD_COUPON_IMPORT_URL' ) ) {
			define( 'EDD_COUPON_IMPORT_URL', plugin_dir_url( __FILE__ ) );
		}
		// plugin folder path
		if ( !defined( 'EDD_COUPON_IMPORT_DIR' ) ) {
			define( 'EDD_COUPON_IMPORT_DIR', plugin_dir_path( __FILE__ ) );
		}
		// plugin root file
		if ( !defined( 'EDD_COUPON_IMPORT_FILE' ) ) {
			define( 'EDD_COUPON_IMPORT_FILE', __FILE__ );
		}
		if( !class_exists( 'EDD_SL_Plugin_Updater' ) ) {
			// load our custom updater
			include( dirname( __FILE__ ) . '/EDD_SL_Plugin_Updater.php' );
		}

		$this->includes();
		$this->init();

	}


	/**
	 * Include our extra files
	 *
	 * @since 1.0
	 *
	 * @access private
	 * @return void
	 */
	private function includes() {

		if( is_admin() ) {
			include dirname( __FILE__ ) . '/includes/admin-actions.php';
			include dirname( __FILE__ ) . '/includes/admin-page.php';
			include dirname( __FILE__ ) . '/includes/import-functions.php';
		}

	}


	/**
	 * Run action and filter hooks.
	 *
	 * @since 1.0
	 *
	 * @access private
	 * @return void
	 */
	private function init() {

		if( ! class_exists( 'Easy_Digital_Downloads' ) )
			return; // EDD not present

		global $edd_options;

		// internationalization
		add_action( 'init', array( $this, 'textdomain' ) );

		// Register the global
		global $edd_import_coupon_file_columns;
        $edd_import_coupon_file_columns = array(
             "0" => "A",
            "1" => "B",
            "2" => "C",
            "3" => "D",
            "4" => "E",
            "5" => "F",
            "6" => "G",
            "7" => "H",
            "8" => "I",
            "9" => "J",
            "10" => "K",
            "11" => "L",
            "12" => "M",
            "13" => "N",
            "14" => "O",
            "15" => "P",
            "16" => "Q",
            "17" => "R",
            "18" => "S",
            "19" => "T",
            "20" => "U",
            "21" => "V",
            "22" => "W",
            "23" => "X",
            "24" => "Y",
            "25" => "Z",
            "26" => "AA",
            "27" => "BB",
            "28" => "CC",
            "29" => "DD",
            "30" => "EE",
            "31" => "FF",
            "32" => "GG",
            "33" => "HH",
            "34" => "II",
            "35" => "JJ",
            "36" => "KK",
            "37" => "LL",
            "38" => "MM",
            "39" => "NN",
            "40" => "OO",
            "41" => "PP",
            "42" => "QQ",
            "43" => "RR",
            "44" => "SS",
            "45" => "TT",
            "46" => "UU",
            "47" => "VV",
            "48" => "WW",
            "49" => "XX",
            "50" => "YY",
            "51" => "ZZ" 
        );

		// register our license key settings
		add_filter( 'edd_settings_misc', array( $this, 'settings' ), 1 );

		// activate license key on settings save
		add_action( 'admin_init', array( $this, 'activate_license' ) );
		add_action( 'admin_init', array( $this, 'deactivate_license' ) );

		// auto updater

		// retrieve our license key from the DB
		$edd_ci_license_key = isset( $edd_options['edd_ci_license_key'] ) ? trim( $edd_options['edd_ci_license_key'] ) : '';

		// setup the updater
		$edd_updater = new EDD_SL_Plugin_Updater( EDD_CI_STORE_API_URL, __FILE__, array(
				'version' 	=> EDD_CI_VERSION, 		// current version number
				'license' 	=> $edd_ci_license_key, // license key (used get_option above to retrieve from DB)
				'item_name' => EDD_CI_PRODUCT_NAME, // name of this plugin
				'author' 	=> 'Chris Christoff'  // author of this plugin
			)
		);

	}


	/**
	 * Load plugin text domain
	 *
	 * @since 1.0
	 *
	 * @access public
	 * @return void
	 */
	public static function textdomain() {

		// Set filter for plugin's languages directory
		$lang_dir = dirname( plugin_basename( __FILE__ ) ) . '/languages/';
		$lang_dir = apply_filters( 'edd_ci_lang_directory', $lang_dir );

		// Load the translations
		load_plugin_textdomain( 'edd-ci', false, $lang_dir );

	}


	/**
	 * Add our extension settings
	 *
	 * @since 1.0
	 * @todo STDS for the columns
	 * @access public
	 * @return array
	 */
	public function settings( $settings ) {
	global $edd_import_coupon_file_columns;
	$ci_settings = array(
	array(
					'id'   => 'edd_ci_settings',
					'name' => '<strong>' . __( 'Coupon Importer', 'edd' ) . '</strong>',
					'desc' => '',
					'type' => 'header',
					'size' => 'regular'
		),

	array(
					'id'   => 'edd_ci_settings_license_key',
					'name' => __('License Key', 'edd'),
					'desc' => __('Enter your license for EDD Coupon Importer to receive automatic upgrades', 'edd'),
					'type' => 'license_key',
					'size' => 'regular',
					'options' => array( 'is_valid_license_option' => 'edd_ci_license_active' )
		),
	array(
					'id' => 'edd_csv_discount_import',
					'name' => '<strong>' . __('CSV Coupon Import Mapping', 'edd') . '</strong>',
					'desc' => __('', 'edd'),
					'type' => 'header',
				),
	array(
					'id' 		=> 'edd_import_discount_name',
					'name'		=> __('Coupon Name','edd'),
					'desc'		=> __('The name of the coupon.', 'edd'),
					'type' 		=> 'select',
					'options'	=> $edd_import_coupon_file_columns
				),
	array(
					'id' 		=> 'edd_import_discount_code',
					'name'		=> __('Coupon Code','edd'),
					'desc'		=> __('The code users will enter to recieve the discount.', 'edd'),
					'type' 		=> 'select',
					'options'	=> $edd_import_coupon_file_columns
				),
	array(
					'id' 		=> 'edd_import_discount_type',
					'name'		=> __('Type of Discount','edd'),
					'desc'		=> __('Possible values: \'flat\' or \'percentage\'', 'edd'),
					'type' 		=> 'select',
					'options'	=> $edd_import_coupon_file_columns
				),
	array(
					'id' 		=> 'edd_import_discount_start',
					'name'		=> __('Date and Time when the discount becomes valid','edd'),
					'desc'		=> __('Format: \'YYYY-MM-DD HH:MM:SS\'', 'edd'),
					'type' 		=> 'select',
					'options'	=> $edd_import_coupon_file_columns
				),
	array(
					'id' 		=> 'edd_import_discount_end',
					'name'		=> __('Date and Time when the discount becomes invalid','edd'),
					'desc'		=> __('Format: \'YYYY-MM-DD HH:MM:SS\'', 'edd'),
					'type' 		=> 'select',
					'options'	=> $edd_import_coupon_file_columns
				),
	array(
					'id' 		=> 'edd_import_discount_status',
					'name'		=> __('Status:','edd'),
					'desc'		=> __('Possible values: \'active\',\'inactive\',or \'expired\'', 'edd'),
					'type' 		=> 'select',
					'options'	=> $edd_import_coupon_file_columns
				),
	array(
					'id' 		=> 'edd_import_coupon_uses',
					'name'		=> __('Current number of uses:','edd'),
					'desc'		=> __('The number of times the coupon has already been used.', 'edd'),
					'type' 		=> 'select',
					'options'	=> $edd_import_coupon_file_columns
				),
	array(
					'id' 		=> 'edd_import_discount_max_uses',
					'name'		=> __('Max uses:','edd'),
					'desc'		=> __('The number of times the coupon may be used.', 'edd'),
					'type' 		=> 'select',
					'options'	=> $edd_import_coupon_file_columns
				),
	array(
					'id' 		=> 'edd_import_discount_amount',
					'name'		=> __('Amount of Discount:','edd'),
					'desc'		=> __('The amount of discount.', 'edd'),
					'type' 		=> 'select',
					'options'	=> $edd_import_coupon_file_columns
				),
	array(
					'id' 		=> 'edd_import_discount_min_price',
					'name'		=> __('Min Price:','edd'),
					'desc'		=> __('The amount that must be in cart before the discount may be used.', 'edd'),
					'type' 		=> 'select',
					'options'	=> $edd_import_coupon_file_columns
				),
	array(
					'id' 		=> 'edd_import_discount_product_reqs',
					'name'		=> __('Product Requirements:','edd'),
					'desc'		=> __('Products that must be in cart for coupon to be used. Valid input: If one product, just the id. If multiple, seperate by pipes: 20|32|108', 'edd'),
					'type' 		=> 'select',
					'options'	=> $edd_import_coupon_file_columns
				),
	array(
					'id' 		=> 'edd_import_discount_product_condition',
					'name'		=> __('Product Condition:','edd'),
					'desc'		=> __('Require \'all\' or \'any\' of the products from the product requirements for discount to be valid. ', 'edd'),
					'type' 		=> 'select',
					'options'	=> $edd_import_coupon_file_columns
				),
	array(
					'id' 		=> 'edd_import_discount_is_not_global',
					'name'		=> __('Is Global:','edd'),
					'desc'		=> __('Apply discount only to products from the product requirements. Either 1 for yes, or 0 for no.', 'edd'),
					'type' 		=> 'select',
					'options'	=> $edd_import_coupon_file_columns
				),
	array(
					'id' 		=> 'edd_import_discount_is_single_use',
					'name'		=> __('Is  Single Use:','edd'),
					'desc'		=> __('Limit this discount to a single-use per customer? Either 1 for yes, or 0 for no.', 'edd'),
					'type' 		=> 'select',
					'options'	=> $edd_import_coupon_file_columns
				),
	array(
					'id' 		=> 'edd_import_discount_id',
					'name'		=> __('Discount ID:','edd'),
					'desc'		=> __('If discount already exists, the ID of it. Else empty string.', 'edd'),
					'type' 		=> 'select',
					'options'	=> $edd_import_coupon_file_columns
				)
				);
	return array_merge( $settings, $ci_settings );
}


	/**
	 * Activate a license key
	 *
	 * @since 1.0
	 *
	 * @access public
	 * @return void
	 */
	public function activate_license() {

		global $edd_options;

		if( ! isset( $_POST['edd_settings_misc'] ) )
			return;
		if( ! isset( $_POST['edd_settings_misc']['edd_ci_license_key'] ) )
			return;

		if( get_option( 'edd_ci_license_active' ) == 'valid' )
			return;

		$license = sanitize_text_field( $_POST['edd_settings_misc']['edd_ci_license_key'] );

		// data to send in our API request
		$api_params = array(
			'edd_action'=> 'activate_license',
			'license' 	=> $license,
			'item_name' => urlencode( EDD_CI_PRODUCT_NAME ) // the name of our product in EDD
		);

		// Call the custom API.
		$response = wp_remote_get( add_query_arg( $api_params, EDD_CI_STORE_API_URL ), array( 'timeout' => 15, 'body' => $api_params, 'sslverify' => false ) );

		// make sure the response came back okay
		if ( is_wp_error( $response ) )
			return false;

		// decode the license data
		$license_data = json_decode( wp_remote_retrieve_body( $response ) );

		update_option( 'edd_ci_license_active', $license_data->license );

	}


	/**
	 * Deactivate a license key
	 *
	 * @since  1.1.1
	 * @return void
	 */

	public function deactivate_license() {
		global $edd_options;

		if ( ! isset( $_POST['edd_settings_misc'] ) )
			return;

		if ( ! isset( $_POST['edd_settings_misc']['edd_ci_license_key'] ) )
			return;

		// listen for our activate button to be clicked
		if( isset( $_POST['edd_ci_license_key_deactivate'] ) ) {

		    // run a quick security check
		    if( ! check_admin_referer( 'edd_ci_license_key_nonce', 'edd_ci_license_key_nonce' ) )
		      return; // get out if we didn't click the Activate button

		    // retrieve the license from the database
		    $license = trim( $edd_options['edd_ci_license_key'] );

		    // data to send in our API request
		    $api_params = array(
		      'edd_action'=> 'deactivate_license',
		      'license'   => $license,
		      'item_name' => urlencode( EDD_CI_PRODUCT_NAME ) // the name of our product in EDD
		    );

		    // Call the custom API.
		    $response = wp_remote_get( add_query_arg( $api_params, EDD_CI_STORE_API_URL ), array( 'timeout' => 15, 'sslverify' => false ) );

		    // make sure the response came back okay
		    if ( is_wp_error( $response ) )
		    	return false;

		    // decode the license data
		    $license_data = json_decode( wp_remote_retrieve_body( $response ) );

		    // $license_data->license will be either "deactivated" or "failed"
		    if( $license_data->license == 'deactivated' )
		    	delete_option( 'edd_ci_license_active' );

		}
	}

}


/**
 * Get everything running
 *
 * @since 1.0
 *
 * @access private
 * @return void
 */

function edd_ci_load() {
	$discounts = new EDD_CI();
}
add_action( 'plugins_loaded', 'edd_ci_load' );



/**
 * Registers the new license field type
 *
 * @access      private
 * @since       10
 * @return      void
*/

if( ! function_exists( 'edd_license_key_callback' ) ) {
	function edd_license_key_callback( $args ) {
		global $edd_options;

		if( isset( $edd_options[ $args['id'] ] ) ) { $value = $edd_options[ $args['id'] ]; } else { $value = isset( $args['std'] ) ? $args['std'] : ''; }
		$size = isset( $args['size'] ) && !is_null($args['size']) ? $args['size'] : 'regular';
		$html = '<input type="text" class="' . $args['size'] . '-text" id="edd_settings_' . $args['section'] . '[' . $args['id'] . ']" name="edd_settings_' . $args['section'] . '[' . $args['id'] . ']" value="' . esc_attr( $value ) . '"/>';

		if( 'valid' == get_option( $args['options']['is_valid_license_option'] ) ) {
			$html .= wp_nonce_field( $args['id'] . '_nonce', $args['id'] . '_nonce', false );
			$html .= '<input type="submit" class="button-secondary" name="' . $args['id'] . '_deactivate" value="' . __( 'Deactivate License',  'edd-recurring' ) . '"/>';
		}
		$html .= '<label for="edd_settings_' . $args['section'] . '[' . $args['id'] . ']"> '  . $args['desc'] . '</label>';

		echo $html;
	}
}